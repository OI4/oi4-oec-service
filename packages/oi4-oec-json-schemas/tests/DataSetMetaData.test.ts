import { matchersWithOptions } from 'jest-json-schema'
import schema from '../src/schemas/DataSetMetaData.schema.json'
import uint16 from '../src/schemas/dataTypes/uint16.schema.json'
import uint32 from '../src/schemas/dataTypes/uint32.schema.json'
import int32 from '../src/schemas/dataTypes/int32.schema.json'
import byte from '../src/schemas/dataTypes/byte.schema.json'
import keyValuePair from '../src/schemas/KeyValuePair.schema.json'
import baseDataType from '../src/schemas/BaseDataType.schema.json'
import enumDescription from '../src/schemas/EnumDescription.schema.json'
import qualifiedName from '../src/schemas/QualifiedName.schema.json'
import localizedTextSchema from '../src/schemas/LocalizedText.schema.json'
import locale from '../src/schemas/constants/locale.pattern.schema.json'
import dataSetMetaDataTypeSchema from '../src/schemas/DataSetMetaDataType.schema.json'
import configurationVersionDataTypeSchema from '../src/schemas/ConfigurationVersionDataType.schema.json'
import fieldMetaDataSchema from '../src/schemas/FieldMetaData.schema.json'
import nodeIdSchema from '../src/schemas/NodeId.schema.json'
import structureDescription from '../src/schemas/StructureDescription.schema.json'
import simpleTypeDescription from '../src/schemas/SimpleTypeDescription.schema.json'
import oi4Identifier from '../src/schemas/oi4Identifier.schema.json'

import validObjs from './__fixtures__/DataSetMetaData_valid.json'
import invalidObjs from './__fixtures__/DataSetMetaData_invalid.json'


expect.extend(
  matchersWithOptions({
    // Loading in a schema which is comprised only of definitions,
    // which means specific test schemas need to be created.
    // This is good for testing specific conditions for definition schemas.
    schemas: [uint16, uint32, int32, byte, localizedTextSchema, locale, 
      keyValuePair, qualifiedName, baseDataType, enumDescription, structureDescription,
    simpleTypeDescription, oi4Identifier],
    verbose: true,
  }, (ajv) => {
    ajv.addSchema(dataSetMetaDataTypeSchema);
    ajv.addSchema(configurationVersionDataTypeSchema);
    ajv.addSchema(fieldMetaDataSchema);
    ajv.addSchema(nodeIdSchema);
  })
)

describe('DataSetMetaData schema', () => {
    it('validate schema', () => {
      expect(schema).toBeValidSchema()
    })
  
    it.each(validObjs as [])(
      '(%#) match valid config object to schema -> %s',
      (_name: string, obj) => {
        expect(obj).toMatchSchema(schema)
      }
    )
  
    it.each(invalidObjs as [])(
      '(%#) match fails for invalid config -> %s',
      (name: string, obj) => {
        expect(obj).not.toMatchSchema(schema)
        expect(() => {
          expect(obj).toMatchSchema(schema)
        }).toThrowErrorMatchingSnapshot(name)
      }
    )
  })