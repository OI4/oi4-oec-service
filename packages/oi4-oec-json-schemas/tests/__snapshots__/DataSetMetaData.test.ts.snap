// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSetMetaData schema (0) match fails for invalid config -> empty: empty 1`] = `
"expect(received).toMatchSchema(schema)

received
  must have required property 'DataSetWriterId'
    Path:     DataSetMetaData.schema.json#/required
  must have required property 'filter'
    Path:     DataSetMetaData.schema.json#/required
  must have required property 'MessageId'
    Path:     DataSetMetaData.schema.json#/required
  must have required property 'MessageType'
    Path:     DataSetMetaData.schema.json#/required
  must have required property 'PublisherId'
    Path:     DataSetMetaData.schema.json#/required
  must have required property 'MetaData'
    Path:     DataSetMetaData.schema.json#/required
  must have required property 'subResource'
    Path:     DataSetMetaData.schema.json#/required
"
`;

exports[`DataSetMetaData schema (1) match fails for invalid config -> Invalid message id: Invalid message id 1`] = `
"expect(received).toMatchSchema(schema)

received
  /MessageId must match pattern \\"^.{1,}-(Registry|OTConnector|Utility|Persistence|Aggregation|OOCConnector|ITConnector)\\\\/(([a-z0-9-]+\\\\.)*([a-z0-9-]*)(?:\\\\/[^\\\\/\`\\\\^
]+){3})$\\"
    Received: <string> provider.com/FieldDataService/FDS-001/1200-0345
    Path:     DataSetMetaData.schema.json#/properties/MessageId/pattern
  /PublisherId must match pattern \\"^(Registry|OTConnector|Utility|Persistence|Aggregation|OOCConnector|ITConnector)\\\\/(([a-z0-9-]+\\\\.)*([a-z0-9-]*)(?:\\\\/[^\\\\/\`\\\\^
]+){3})$\\"
    Received: <string> provider.com/FieldDataService/FDS-001/1200-0345
    Path:     DataSetMetaData.schema.json#/properties/PublisherId/pattern
"
`;

exports[`DataSetMetaData schema (2) match fails for invalid config -> Invalid publisher id: Invalid publisher id 1`] = `
"expect(received).toMatchSchema(schema)

received
  /PublisherId must match pattern \\"^(Registry|OTConnector|Utility|Persistence|Aggregation|OOCConnector|ITConnector)\\\\/(([a-z0-9-]+\\\\.)*([a-z0-9-]*)(?:\\\\/[^\\\\/\`\\\\^
]+){3})$\\"
    Received: <string> provider.com
    Path:     DataSetMetaData.schema.json#/properties/PublisherId/pattern
"
`;
